program/rust/src/time_machine_types.rs

```diff
@@ -6,20 +6,371 @@ use crate::c_oracle_header::{
+
+    ///add a new price to the tracker
+    fn add_price(
+        &mut self,
+        prev_time: i64,
+        prev_price: i64,
+        prev_conf: u64,
+        current_time: i64,
+        current_price: i64,
+        current_conf: u64,
+    ) -> Result<(), OracleError>;
+    /// returns the minimum number of entries that needs to be added
+    /// to the current entry of the time_machine so that the following two
+    /// conditions are satisfied
+    /// 1. after adding the output, the new current entry has the same offset as if
+    ///     the price was constantly updating
+    /// 2. any entry that would have been passed through during this time interval, would be
+    ///     passed through if we skip one entry at a time the number that this outputs
+    fn get_num_skipped_entries_capped(
+        prev_time: i64,
+        current_time: i64,
+    ) -> Result<usize, OracleError> {
+        let time_since_begining_of_current_entry =
+            (prev_time % GRANUALITY) + current_time - prev_time;
+        let mut num_skiped_entries: usize =
+            try_convert(time_since_begining_of_current_entry / GRANUALITY)?;
+        //if num_skipped is greater than NUM_ENTRIES, cap it at the smallest
+        //number greater than NUM_ENTRIES that has the same residue class
+        if num_skiped_entries >= NUM_ENTRIES {
+            num_skiped_entries %= NUM_ENTRIES;
+            num_skiped_entries += NUM_ENTRIES;
+        }
+        Ok(num_skiped_entries)
+    }
+    ///gets the index of the next entry
+    fn get_next_entry(current_entry: usize) -> usize {
+        (current_entry + 1) % NUM_ENTRIES
+    }
+
+    ///returns the remining time before the next multiple of GRANUALITY
+    fn get_time_to_entry_end(time: i64) -> i64 {
+        (GRANUALITY - (time % GRANUALITY)) % GRANUALITY
+    }
+}
+
+#[derive(Debug, Copy, Clone)]
+#[repr(C)]
+/// Represents an SMA Tracker that has NUM_ENTRIES entries
+/// each tracking time weighted sums for GRANUALITY seconds periods.
+///The prices are assumed to be provided under some fixed point representation, and the computation
+/// gurantees accuracy up to the last decimal digit in the fixed point representation.
+/// Assumes THRESHOLD < GRANUALITY
+pub struct SmaTracker<const GRANUALITY: i64, const NUM_ENTRIES: usize, const THRESHOLD: i64> {
+    valid_entry_counter:        u64, // is incremented everytime we add a valid entry
+    max_update_time:            i64, // maximum time between two updates in the current entry.
+    running_price:              [SignedTrackerRunningSum; NUM_ENTRIES], /* price running time
+                                      * weighted sums */
+    running_confidence:         [UnsignedTrackerRunningSum; NUM_ENTRIES], /* confidence running
+                                                                           * time
+                                                                           * weighted sums */
+    current_entry:              usize, //the current entry
```

Comment:
I'm not sure you need to store this, because it's just `current_time / GRANULARTIY`

Reply:
At the moment it's not exactly that. But yes, we can change things so that it is  ((current_time / GRANULARTIY) % NUM_ENTRIES). Which sounds like a good idea